telemetry/tracks/class-tracks-event.php

```diff
@@ -0,0 +1,336 @@
+		if ( defined( 'VIP_GO_APP_ENVIRONMENT' ) ) {
+			$app_environment = constant( 'VIP_GO_APP_ENVIRONMENT' );
+			if ( is_string( $app_environment ) && '' !== $app_environment ) {
+				$event->vipgo_env = $app_environment;
+			}
+		}
+
+		// Set VIP organization if it exists.
+		if ( defined( 'VIP_ORG_ID' ) ) {
+			$org_id = constant( 'VIP_ORG_ID' );
+			if ( is_string( $org_id ) && '' !== $org_id ) {
+				$event->vipgo_org = $org_id;
+			}
+		}
+
+		// Check if the user is a VIP user.
+		$event->is_vip_user = Support_User::user_has_vip_support_role( get_current_user_id() );
+
+		return $event;
+	}
+
+	/**
+	 * Sets the Tracks User ID and User ID Type depending on the current
+	 * environment.
+	 *
+	 * @param Tracks_Event_DTO $event The event to annotate with identity information.
+	 * @return Tracks_Event_DTO The new event object including identity information.
+	 */
+	protected static function set_user_properties( Tracks_Event_DTO $event ): Tracks_Event_DTO {
+		$wp_user = wp_get_current_user();
+
+		// Only track logged-in users.
+		if ( 0 === $wp_user->ID ) {
+			return $event;
+		}
+
+		// Set anonymized event user ID; it should be consistent across environments.
+		if ( defined( 'VIP_TELEMETRY_SALT' ) ) {
+			$salt           = constant( 'VIP_TELEMETRY_SALT' ); // this would be private but globally shared across the platform
+			$tracks_user_id = hash_hmac( 'sha256', $wp_user->user_email, $salt );
+
+			$event->_ui = $tracks_user_id;
+			$event->_ut = 'vip:user_email';
+
+			return $event;
+		}
+
+		// Users in the VIP environment.
+		if ( defined( 'VIP_GO_APP_ID' ) ) {
+			$app_id = constant( 'VIP_GO_APP_ID' );
+			if ( is_integer( $app_id ) && 0 < $app_id ) {
+				$event->_ui = sprintf( '%s_%s', $app_id, $wp_user->ID );
+				$event->_ut = 'vip_go_app_wp_user';
+
+				return $event;
+			}
+		}
+
+		// All other environments.
+		$wp_base_url = get_option( 'home' );
+		if ( ! is_string( $wp_base_url ) || '' === $wp_base_url ) {
+			$wp_base_url = get_option( 'siteurl' );
+		}
+
+		/**
+		 * The base URL of the site.
+		 *
+		 * @var string $wp_base_url
+		 */
+		$event->_ui = wp_hash( sprintf( '%s|%s', $wp_base_url, $wp_user->ID ) );
+		$event->_ut = 'anon'; // Same as the default value in the original code.
+
+		return $event;
+	}
+
+	/**
+	 * Validates the event object.
+	 *
+	 * @param Tracks_Event_DTO $event Event object to validate.
+	 * @return ?WP_Error null if validation passed, error otherwise.
+	 */
+	protected static function get_event_validation_result( Tracks_Event_DTO $event ): ?WP_Error {
+		// Check that required fields are defined.
+		if ( ! $event->_en ) {
+			$msg = __( 'The _en property must be specified to non-empty value', 'vip-telemetry' );
+			log2logstash( [
+				'severity' => 'error',
+				'feature'  => 'telemetry',
+				'message'  => $msg,
+				'extra'    => [
+					'event' => (array) $event,
+				],
+			] );
+			return new WP_Error(
+				'invalid_event',
+				$msg,
+				array( 'status' => 400 )
```

Comment:
Do these statuses have a functional use?

Reply:
I don't recall. I left them in because there's no harm, but if you think otherwise, let's cut them out.

Reply:
I usually see these in the context of a REST response, but in here they're all consumed by other code. I think they should be remove unless we're actually using these for HTTP status responses, as they imply something misleading.

Reply:
Makes sense, they are gone in f70cc5543